File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/service/CostCalculationService.java
package com.study.webflux.rag.domain.service;

import com.study.webflux.rag.domain.model.cost.CostInfo;
import com.study.webflux.rag.domain.model.cost.ModelPricing;
import com.study.webflux.rag.domain.model.metrics.UsageAnalytics;

public class CostCalculationService {

	public static CostInfo calculateCost(UsageAnalytics analytics) {
		if (analytics == null) {
			return CostInfo.zero();
		}

		long llmCredits = calculateLlmCredits(analytics);
		long ttsCredits = calculateTtsCredits(analytics);

		return CostInfo.of(llmCredits, ttsCredits);
	}

	private static long calculateLlmCredits(UsageAnalytics analytics) {
		if (analytics.llmUsage() == null) {
			return 0;
		}

		String model = analytics.llmUsage().model();
		int tokenCount = analytics.llmUsage().tokenCount();

		int estimatedPromptTokens = estimatePromptTokens(analytics);
		int completionTokens = tokenCount;

		return ModelPricing.calculateLlmCredits(model, estimatedPromptTokens, completionTokens);
	}

	private static int estimatePromptTokens(UsageAnalytics analytics) {
		if (analytics.userRequest() == null) {
			return 0;
		}

		int inputLength = analytics.userRequest().inputLength();
		int memoryCount = analytics.retrievalMetrics() != null
			? analytics.retrievalMetrics().memoryCount()
			: 0;
		int documentCount = analytics.retrievalMetrics() != null
			? analytics.retrievalMetrics().documentCount()
			: 0;

		int basePromptTokens = 300;
		int inputTokens = inputLength / 3;
		int contextTokens = (memoryCount * 50) + (documentCount * 100);

		return basePromptTokens + inputTokens + contextTokens;
	}

	private static long calculateTtsCredits(UsageAnalytics analytics) {
		if (analytics.ttsMetrics() == null) {
			return 0;
		}

		int sentenceCount = analytics.ttsMetrics().sentenceCount();

		long estimatedAudioLength = estimateAudioLength(sentenceCount);

		return ModelPricing.calculateTtsCredits(estimatedAudioLength);
	}

	private static long estimateAudioLength(int sentenceCount) {
		long avgSentenceDuration = 3000;
		return sentenceCount * avgSentenceDuration;
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/cost/service/CostCalculationService.kt
package com.miyou.app.domain.cost.service

import com.miyou.app.domain.cost.model.CostInfo
import com.miyou.app.domain.cost.model.ModelPricing
import com.miyou.app.domain.monitoring.model.UsageAnalytics

class CostCalculationService {
    companion object {
        @JvmStatic
        fun calculateCost(analytics: UsageAnalytics?): CostInfo {
            if (analytics == null) {
                return CostInfo.zero()
            }

            val llmCredits = calculateLlmCredits(analytics)
            val ttsCredits = calculateTtsCredits(analytics)

            return CostInfo.of(llmCredits, ttsCredits)
        }

        private fun calculateLlmCredits(analytics: UsageAnalytics): Long {
            val llmUsage = analytics.llmUsage ?: return 0L
            val model = llmUsage.model

            val promptTokens =
                llmUsage.promptTokens.takeIf { it != null }
                    ?: estimatePromptTokens(analytics)
            val completionTokens = llmUsage.completionTokens ?: llmUsage.totalTokens

            return ModelPricing.calculateLlmCredits(model, promptTokens, completionTokens)
        }

        private fun estimatePromptTokens(analytics: UsageAnalytics): Int {
            val userRequest = analytics.userRequest ?: return 0

            val inputLength = userRequest.inputLength
            val retrievalMetrics = analytics.retrievalMetrics
            val memoryCount = retrievalMetrics?.memoryCount ?: 0
            val documentCount = retrievalMetrics?.documentCount ?: 0

            val basePromptTokens = 300
            val inputTokens = inputLength / 3
            val contextTokens = (memoryCount * 50) + (documentCount * 100)

            return basePromptTokens + inputTokens + contextTokens
        }

        private fun calculateTtsCredits(analytics: UsageAnalytics): Long {
            val ttsMetrics = analytics.ttsMetrics ?: return 0L
            val sentenceCount = ttsMetrics.sentenceCount
            val estimatedAudioLength = estimateAudioLength(sentenceCount)

            return ModelPricing.calculateTtsCredits(estimatedAudioLength)
        }

        private fun estimateAudioLength(sentenceCount: Int): Long {
            val avgSentenceDuration = 3000L
            return sentenceCount.toLong() * avgSentenceDuration
        }
    }
}
